File: src/go/PhantasyIslandGoRemoteControl/radio/radio_manager.go
package radio

import (
	"context"
	"fmt"
	"math"
	"sync"
	"time"

	socketio "github.com/googollee/go-socket.io"
)

// WaiterRef 模拟弱引用。在 Go 中没有直接的弱引用，
// 我们通过在 Token 完成或超时后手动从 map 中移除来模拟。
type WaiterRef struct {
	token *WaitToken
}

type RadioManager struct {
	client      *socketio.Client // 这里假设使用某个 go-socket.io 客户端库
	Namespace   string
	SceneIsInit bool

	// Sub-API 模块
	DebugApi *DebugApi
	SceneApi *SceneApi
	FlyApi   *FlyApi
	RadioApi *RadioApi

	pendingWaiters map[string][]*WaitToken
	waitersLock    sync.Mutex
}

func NewRadioManager() *RadioManager {
	rm := &RadioManager{
		Namespace:      "/UserSide",
		SceneIsInit:    false,
		pendingWaiters: make(map[string][]*WaitToken),
	}
	rm.DebugApi = NewDebugApi(rm)
	rm.SceneApi = NewSceneApi(rm)
	rm.FlyApi = NewFlyApi(rm)
	rm.RadioApi = NewRadioApi(rm)
	return rm
}

func (rm *RadioManager) CreateMsgTimestampID() int64 {
	return int64(math.Floor(float64(time.Now().UnixNano()) / 1e4))
}

func (rm *RadioManager) Connect(url string, namespace string) error {
	// 实际连接逻辑取决于所选的 socket.io 库
	rm.Namespace = namespace
	rm.initListener()
	fmt.Printf("[RadioManager] Connecting to %s%s\n", url, namespace)
	return nil
}

func (rm *RadioManager) initListener() {
	// 注册各种事件回调
}

func (rm *RadioManager) Ping() interface{} {
	return rm.SendAndWaitSync("ping", nil, "pong", 3*time.Second, nil)
}

func (rm *RadioManager) Send(cmd string, data map[string]interface{}) {
	msg := make(map[string]interface{})
	msg["cmd"] = cmd
	for k, v := range data {
		msg[k] = v
	}
	// 执行发送逻辑
}

func (rm *RadioManager) SendWithToken(cmd string, data map[string]interface{}, waitCmd string, postProcessor func(map[string]interface{}) interface{}) *WaitToken {
	if waitCmd == "" {
		waitCmd = cmd
	}

	timeBaseID := rm.CreateMsgTimestampID()
	token := NewWaitToken(waitCmd, timeBaseID)

	if postProcessor != nil {
		token.SetPostProcessor(postProcessor)
	}

	rm.waitersLock.Lock()
	rm.pendingWaiters[waitCmd] = append(rm.pendingWaiters[waitCmd], token)
	rm.waitersLock.Unlock()

	msg := make(map[string]interface{})
	msg["timestampIdPython"] = timeBaseID
	for k, v := range data {
		if v != nil {
			msg[k] = v
		}
	}

	rm.Send(cmd, msg)
	return token
}

func (rm *RadioManager) SendAndWaitSync(cmd string, data map[string]interface{}, waitCmd string, timeout time.Duration, postProcessor func(map[string]interface{}) interface{}) interface{} {
	token := rm.SendWithToken(cmd, data, waitCmd, postProcessor)
	return token.Wait(timeout)
}

func (rm *RadioManager) SendAndWaitAsync(ctx context.Context, cmd string, data map[string]interface{}, waitCmd string, postProcessor func(map[string]interface{}) interface{}) (interface{}, error) {
	token := rm.SendWithToken(cmd, data, waitCmd, postProcessor)
	return token.WaitAsync(ctx)
}

func (rm *RadioManager) NotifyWaiters(cmd string, data map[string]interface{}) bool {
	timestampIDRaw, ok := data["timestampIdPython"]
	if !ok {
		return false
	}
	timestampID := int64(timestampIDRaw.(float64))

	rm.waitersLock.Lock()
	defer rm.waitersLock.Unlock()

	waiters := rm.pendingWaiters[cmd]
	if len(waiters) == 0 {
		return false
	}

	var surviving []*WaitToken
	matched := false
	for _, token := range waiters {
		if !matched && token.TimeBaseID == timestampID {
			token.Complete(data)
			matched = true
		} else {
			surviving = append(surviving, token)
		}
	}

	if len(surviving) > 0 {
		rm.pendingWaiters[cmd] = surviving
	} else {
		delete(rm.pendingWaiters, cmd)
	}

	return matched
}

func (rm *RadioManager) MsgDispatch(data map[string]interface{}) {
	cmd, _ := data["cmd"].(string)
	if rm.NotifyWaiters(cmd, data) {
		return
	}

	switch cmd {
	case "pong":
	case "sceneReset", "sceneNotInit":
		rm.onSceneReset(data)
	case "sceneInit", "sceneIsInit":
		rm.onSceneInit(data)
	default:
		fmt.Printf("[RadioManager] unknown cmd: %s, data: %v\n", cmd, data)
	}
}

func (rm *RadioManager) onSceneReset(data map[string]interface{}) {
	rm.SceneIsInit = false
}

func (rm *RadioManager) onSceneInit(data map[string]interface{}) {
	rm.SceneIsInit = true
}

File: src/fs/RadioManager.fs
namespace PhantasyIslandRemoteControl.Radio

open System
open System.Collections.Concurrent
open System.Collections.Generic
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open SocketIOClient

type RadioManager() =
    let mutable _socket: SocketIO = null
    let mutable _namespace = "/UserSide"
    let mutable _isSceneInit = false
    
    let _pendingWaiters = ConcurrentDictionary<string, List<WeakReference<WaitToken>>>()
    let _waitersLock = obj()

    interface IRadioManager with
        member this.InternalSend(cmd, ?data) = this.InternalSend(cmd, ?data = data)
        member this.SendWithToken(cmd, ?data, ?waitCmd, ?postProcessor) = this.SendWithToken(cmd, ?data = data, ?waitCmd = waitCmd, ?postProcessor = postProcessor)
        member this.SendAndWaitSync(cmd, ?data, ?waitCmd, ?timeout, ?postProcessor) = this.SendAndWaitSync(cmd, ?data = data, ?waitCmd = waitCmd, ?timeout = timeout, ?postProcessor = postProcessor)
        member this.SendAndWaitAsync(cmd, ?data, ?waitCmd, ?timeout, ?postProcessor) = this.SendAndWaitAsync(cmd, ?data = data, ?waitCmd = waitCmd, ?timeout = timeout, ?postProcessor = postProcessor)

    member this.Socket with get() = _socket and set(v) = _socket <- v
    member this.Namespace with get() = _namespace and set(v) = _namespace <- v
    member this.IsSceneInit with get() = _isSceneInit and set(v) = _isSceneInit <- v

    // API Properties
    member this.DebugApi = DebugApi(this :> IRadioManager)
    member this.SceneApi = SceneApi(this :> IRadioManager)
    member this.FlyApi = FlyApi(this :> IRadioManager)
    member this.RadioApi = RadioApi(this :> IRadioManager)

    member this.CreateMsgTimestampId() =
        DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() * 100L

    member this.Connect(?url: string, ?ns: string) =
        async {
            let urlVal = defaultArg url "http://127.0.0.1:60002"
            this.Namespace <- defaultArg ns "/UserSide"
            this.Reset()
            
            _socket <- new SocketIO(urlVal)
            this.InitListener()
            _socket.ConnectAsync() |> Async.AwaitTask |> ignore
        }

    member this.Reset() =
        if _socket <> null && _socket.Connected then
            _socket.DisconnectAsync().Wait()
        _isSceneInit <- false

    member private this.InitListener() =
        if _socket = null then ()
        else
            _socket.OnConnected.Add(fun _ ->
                printfn "[RadioManager] connected"
                this.CheckSceneStatus())

            _socket.OnDisconnected.Add(fun _ ->
                printfn "[RadioManager] disconnected"
                _isSceneInit <- false)

            _socket.On("message", fun response ->
                let data = response.GetValue<JsonElement>()
                printfn "[RadioManager] message: %s" (data.ToString())
                this.MsgDispatch(data) |> ignore)

    member private this.CheckSceneStatus() =
        this.InternalSend("ping")
        this.InternalSend("scene.getInitState")

    member this.Ping() =
        this.SendAndWaitSync("ping", waitCmd = "pong")

    member private this.OnSceneReset(data: JsonElement) =
        printfn "[RadioManager] handle sceneReset: %O" data
        _isSceneInit <- false

    member private this.OnSceneInit(data: JsonElement) =
        printfn "[RadioManager] handle sceneInit: %O" data
        _isSceneInit <- true

    member this.InternalSend(cmd: string, ?data: Dictionary<string, obj>) =
        let msg = Dictionary<string, obj>()
        msg.["cmd"] <- cmd
        match data with
        | Some d -> for kv in d do msg.[kv.Key] <- kv.Value
        | None -> ()
        
        if _socket <> null then
            _socket.EmitAsync("message", msg) |> ignore
        ()

    member this.SendWithToken(cmd: string, ?data: Dictionary<string, obj>, ?waitCmd: string, ?postProcessor: Func<JsonElement, obj>) =
        let wCmd = defaultArg waitCmd cmd
        let timeBaseId = this.CreateMsgTimestampId()
        let token = WaitToken(wCmd, timeBaseId)
        match postProcessor with
        | Some pp -> token.SetPostProcessor(pp) |> ignore
        | None -> ()

        lock _waitersLock (fun () ->
            let waiters = _pendingWaiters.GetOrAdd(wCmd, fun _ -> List<WeakReference<WaitToken>>())
            waiters.Add(WeakReference<WaitToken>(token)))

        let msg = Dictionary<string, obj>()
        msg.["timestampIdPython"] <- timeBaseId
        match data with
        | Some d -> 
            for kv in d do 
                if kv.Value <> null then msg.[kv.Key] <- kv.Value
        | None -> ()

        this.InternalSend(cmd, msg)
        token

    member this.SendAndWaitSync(cmd: string, ?data: Dictionary<string, obj>, ?waitCmd: string, ?timeout: double, ?postProcessor: Func<JsonElement, obj>) =
        let t = defaultArg timeout 3.0
        let token = this.SendWithToken(cmd, ?data = data, ?waitCmd = waitCmd, ?postProcessor = postProcessor)
        token.Wait(TimeSpan.FromSeconds(t))

    member this.SendAndWaitAsync(cmd: string, ?data: Dictionary<string, obj>, ?waitCmd: string, ?timeout: double, ?postProcessor: Func<JsonElement, obj>) =
        async {
            let t = defaultArg timeout 3.0
            let token = this.SendWithToken(cmd, ?data = data, ?waitCmd = waitCmd, ?postProcessor = postProcessor)
            let! taskResult = token.WaitAsync() |> Async.AwaitTask
            return taskResult
        }

    member private this.NotifyWaiters(cmd: string, data: JsonElement) =
        if String.IsNullOrEmpty cmd then false
        else
            let mutable tsProp = JsonElement()
            if not (data.TryGetProperty("timestampIdPython", &tsProp)) then false
            else
                let timestampId = tsProp.GetInt64()
                lock _waitersLock (fun () ->
                    match _pendingWaiters.TryGetValue(cmd) with
                    | (true, refs) ->
                        let surviving = List<WeakReference<WaitToken>>()
                        let mutable matched = false
                        for r in refs do
                            match r.TryGetTarget() with
                            | (true, token) ->
                                if not matched && token.TimeBaseId = timestampId then
                                    token.Complete(data)
                                    matched <- true
                                else
                                    surviving.Add(r)
                            | _ -> ()
                        
                        if surviving.Count > 0 then _pendingWaiters.[cmd] <- surviving
                        else _pendingWaiters.TryRemove(cmd) |> ignore
                        
                        matched
                    | _ -> false)

    member private this.MsgDispatch(data: JsonElement) : Task =
        let mutable cmdProp = JsonElement()
        if not (data.TryGetProperty("cmd", &cmdProp)) then
            printfn "[RadioManager] received message without cmd: %O" data
            Task.CompletedTask
        else
            let cmd = cmdProp.GetString()
            if not (this.NotifyWaiters(cmd, data)) then
                match cmd with
                | "pong" -> ()
                | "sceneReset" | "sceneNotInit" -> this.OnSceneReset(data)
                | "sceneInit" | "sceneIsInit" -> this.OnSceneInit(data)
                | _ -> printfn "[RadioManager] unknown cmd: %s, data: %O" cmd data
            Task.CompletedTask
